orchagent/dash/dashrouteorch.cpp

```diff
@@ -0,0 +1,994 @@
+#include <cassert>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <exception>
+#include <inttypes.h>
+#include <algorithm>
+
+#include "converter.h"
+#include "dashrouteorch.h"
+#include "macaddress.h"
+#include "orch.h"
+#include "sai.h"
+#include "saiextensions.h"
+#include "swssnet.h"
+#include "tokenize.h"
+
+using namespace std;
+using namespace swss;
+
+extern std::unordered_map<std::string, sai_object_id_t> gVnetNameToId;
+extern sai_dash_eni_api_t* sai_dash_eni_api;
+extern sai_dash_outbound_routing_api_t* sai_dash_outbound_routing_api;
+extern sai_dash_inbound_routing_api_t* sai_dash_inbound_routing_api;
+extern sai_object_id_t gSwitchId;
+extern size_t gMaxBulkSize;
+
+static std::unordered_map<std::string, sai_outbound_routing_entry_action_t> sOutboundAction =
+{
+    { "vnet", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET },
+    { "vnet_direct", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET_DIRECT },
+    { "route_direct", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_DIRECT },
+    { "drop", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_DROP }
+};
+
+DashRouteOrch::DashRouteOrch(DBConnector *db, vector<string> &tableName) :
+    outbound_routing_bulker_(sai_dash_outbound_routing_api, gMaxBulkSize),
+    inbound_routing_bulker_(sai_dash_inbound_routing_api, gMaxBulkSize),
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+}
+
+bool DashRouteOrch::addEniEntry(const string& eni)
+{
+    SWSS_LOG_ENTER();
+
+    if (eni_entries_[eni].eniEntryAdded)
+    {
+        SWSS_LOG_ERROR("ENI entry already exists for %s", eni.c_str());
+        return true;
+    }
+
+    EniEntry &entry = eni_entries_[eni].eni_entry;
+    const string &vnet = entry.vnet;
+
+    if (!vnet.empty() && gVnetNameToId.find(vnet) == gVnetNameToId.end())
+    {
+        SWSS_LOG_ERROR("Retry as vnet %s not found", vnet.c_str());
+        return false;
+    }
+
+    sai_object_id_t &eni_id = entry.eni_id;
+    sai_attribute_t eni_attr;
+    vector<sai_attribute_t> eni_attrs;
+
+    eni_attr.id = SAI_ENI_ATTR_VNET_ID;
+    eni_attr.value.oid = gVnetNameToId[entry.vnet];
+    eni_attrs.push_back(eni_attr);
+
+    bool has_qos = qos_entries_.find(entry.qos_name) != qos_entries_.end();
+
+    eni_attr.id = SAI_ENI_ATTR_PPS;
+    eni_attr.value.u32 = has_qos ? qos_entries_[entry.qos_name].bw : 0;
```

Comment:
The configuration should be set only if it is specified in the config file.
SONiC should not override the vendor's default configuration.
This is applicable to all attributes (SAI_ENI_ATTR_PPS, SAI_ENI_ATTR_CPS, SAI_ENI_ATTR_FLOWS).

Reply:
IMHO, the default value of an SAI attribution should be unified to different vendors if the default value was declared in SAI. 
In the other words, if the default value wasn't specified by SWSS(upper layer of SONiC), the implementation of vendor should follow the SAI definition. e.g. the default value of SAI_ENI_ATTR_PPS is `0` in SAI.(https://github.com/opencomputeproject/SAI/blob/dash/experimental/saiexperimentaldasheni.h#L113)
So, I think to explicitly specified the default value in the upper layer is also acceptable if there is a default value in SAI.

Reply:
@oleksandrivantsiv, I tend to agree with @Pterosaur . The default value of 0 was considered as that was the value specified in SAI headers. Please let us know. We can discuss offline also on this if needed.

Reply:
I'm a little confused here - since the default value is declared in the SAI headers, is it still necessary for orchagent to pass a default value to the SAI API? My understanding is that if we don't provide this attribute the API should automatically fill it with the default value.

@prsunny for other SAI attributes with default values, is it preferred to still program those defaults from orchagent, or let the implementation take care of it?

Reply:
Agree with @theasianpianist , we don't have to provide default from orchagent

Reply:
Looks like at least for neighorch, we are allowing the SAI implementation to handle default values as specified in the SAI headers. For `SAI_NEIGHBOR_ENTRY_ATTR_PACKET_ACTION`, [default is defined as `SAI_PACKET_ACTION_FORWARD`](https://github.com/opencomputeproject/SAI/blob/master/inc/saineighbor.h#L65) and it doesn't look like we set a value for it when [adding a new neighbor](https://github.com/sonic-net/sonic-swss/blob/ab299201be9d37654e3ba88f9e7e180023896722/orchagent/neighorch.cpp#L798). Similarly for `SAI_NEIGHBOR_ENTRY_ATTR_META_DATA` [with a default of 0](https://github.com/opencomputeproject/SAI/blob/master/inc/saineighbor.h#L102), we allow the SAI implementation to set the default and do not use it when creating a new neighbor entry.

Reply:
I will make the change to not set these values explicitly unless there is a specific qos configuration.